public/app/features/library-panels/components/VarImportModal/VarImportModal.tsx

```diff
@@ -0,0 +1,223 @@
+import { DashboardModel } from 'app/features/dashboard/state';
+
+export interface VarImportModalProps {
+  panelVars: VariableModel[];
+  dashboard: DashboardModel;
+  onSubmit: (newVars: VariableModel[]) => void;
+  onDismiss?: () => void;
+  isOpen?: boolean;
+}
+
+const varDef = (v?: QueryVariableModel) => {
+  if (v === undefined) {
+    return '';
+  }
+
+  return v.definition ? v.definition : typeof v.query === 'string' ? v.query : '';
+};
+
+export const VarImportModal: FC<VarImportModalProps> = ({ panelVars, dashboard, onDismiss, onSubmit, isOpen }) => {
+  const styles = useStyles(getStyles);
+  const searchSrv = useMemo(() => new SearchSrv(), []);
+  const [selectedDash, setSelectedDash] = useState<undefined | SelectableValue<string>>(undefined);
+  const [, searchDashboards] = useAsyncFn(
+    async (query = ''): Promise<Array<SelectableValue<string>>> => {
+      const res = await searchSrv.search({
+        query,
+        type: DashboardSearchItemType.DashDB,
+        skipRecent: true,
+        skipStarred: true,
+      });
+      return res?.[0].items
+        .map((v: any) => ({ value: v.uid, label: v.title }))
+        .filter((v: any) => v.value !== dashboard.uid);
+    }
+  );
+
+  const [clashingVars, safeVars] = useMemo(() => {
+    const dashVars = dashboard.templating.list as VariableModel[];
+    const dashVarMap = dashVars.reduce((acc, cur) => {
+      acc[cur.name] = cur;
+      return acc;
+    }, {} as Record<string, VariableModel>);
+
+    const clashes = [];
+    const safe = [];
+    for (const panelVar of panelVars) {
+      if (dashVarMap[panelVar.name] !== undefined) {
+        clashes.push(dashVarMap[panelVar.name]);
+      } else {
+        safe.push(panelVar);
+      }
+    }
+    return [clashes, safe];
+  }, [panelVars]);
+  const [checkedVars, setCheckedVars] = useState<boolean[]>(clashingVars.map(() => false));
+
+  const [dashVars, fetchDashVars] = useAsyncFn(async (dashUid: string) => {
+    const dash = await backendSrv.getDashboardByUid(dashUid);
+    const dashModel = dash.dashboard as DashboardModel;
+    const sourceVars = dashModel.templating.list as VariableModel[];
+    const sourceVarsMap = sourceVars.reduce((acc, cur) => {
+      acc[cur.name] = cur;
+      return acc;
+    }, {} as Record<string, VariableModel>);
+    return sourceVarsMap;
+  }, []);
+
+  const onDashboardSelect = useCallback(async (selectedDash: SelectableValue<string>) => {
+    setSelectedDash(selectedDash);
+    await fetchDashVars(selectedDash.value!);
+  }, []);
+
+  const submitNewVars = useCallback(() => {
+    const selected = clashingVars.filter((v, i) => checkedVars[i]);
+    const newVars = [...safeVars, ...selected].map((v) => dashVars.value?.[v.name]).filter((v) => v !== undefined);
+    onSubmit(newVars as VariableModel[]);
+  }, [dashVars.value, checkedVars, clashingVars, safeVars]);
+
+  const selectedCount = checkedVars.reduce((total, cur) => total + Number(cur), 0);
```

Comment:
I think this might differ between engineers, but I think this component is too complex and has too many responsibilities to a point that it's hard to follow all the hooks and their dependencies. We should look into simplifying this in followup PRs.